### ISSUE TITLE
Home and End keys not working over SSH

### ISSUE THREAD

**Issue by mqudsi** [neovim/neovim#6014]
- `nvim --version`: 

> NVIM 0.1.7
> Build type: Release
> Compilation: /usr/bin/cc -O2 -pipe -DLIBICONV_PLUG -fstack-protector -fno-strict-aliasing -Wconversion -O2 -pipe -DLIBICONV_PLUG -fstack-protector -fno-strict-aliasing -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wno-c11-extensions -Wvla -fstack-protector --param ssp-buffer-size=4 -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/wrkdirs/usr/ports/editors/neovim/work/.build/config -I/wrkdirs/usr/ports/editors/neovim/work/neovim-0.1.7/src -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/include -I/wrkdirs/usr/ports/editors/neovim/work/.build/src/nvim/auto -I/wrkdirs/usr/ports/editors/neovim/work/.build/include
> Compiled by root@103i386-default-job-17
> 
> Optional features included (+) or not (-): +acl   +iconv    -jemalloc +tui
> For differences from Vim, see :help vim-differences
> 
>    system vimrc file: "$VIM/sysinit.vim"
>   fall-back for $VIM: "/usr/local/share/nvim"

- Vim (version: ) behaves differently?
Fully working on

> VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Aug  2 2016 06:09:55)
> Included patches: 1-1832
> Compiled by root@101i386-default-job-18
> Huge version with GTK2 GUI.  Features included (+) or not (-):
> +acl             +farsi           +mouse_netterm   +tag_binary
> +arabic          +file_in_path    +mouse_sgr       +tag_old_static
> +autocmd         +find_in_path    +mouse_sysmouse  -tag_any_white
> +balloon_eval    +float           +mouse_urxvt     +tcl
> +browse          +folding         +mouse_xterm     +termguicolors
> ++builtin_terms  -footer          +multi_byte      +terminfo
> +byte_offset     +fork()          +multi_lang      +termresponse
> +channel         +gettext         -mzscheme        +textobjects
> +cindent         -hangul_input    +netbeans_intg   +timers
> +clientserver    +iconv           +packages        +title
> +clipboard       +insert_expand   +path_extra      +toolbar
> +cmdline_compl   +job             +perl            +user_commands
> +cmdline_hist    +jumplist        +persistent_undo +vertsplit
> +cmdline_info    +keymap          +postscript      +virtualedit
> +comments        +langmap         +printer         +visual
> +conceal         +libcall         +profile         +visualextra
> +cryptv          +linebreak       +python          +viminfo
> +cscope          +lispindent      -python3         +vreplace
> +cursorbind      +listcmds        +quickfix        +wildignore
> +cursorshape     +localmap        +reltime         +wildmenu
> +dialog_con_gui  +lua/dyn         +rightleft       +windows
> +diff            +menu            +ruby            +writebackup
> +digraphs        +mksession       +scrollbind      +X11
> +dnd             +modify_fname    +signs           -xfontset
> -ebcdic          +mouse           +smartindent     +xim
> +emacs_tags      +mouseshape      +startuptime     +xsmp_interact
> +eval            +mouse_dec       +statusline      +xterm_clipboard
> +ex_extra        -mouse_gpm       -sun_workshop    -xterm_save
> +extra_search    -mouse_jsbterm   +syntax          +xpm
>    system vimrc file: "/usr/local/etc/vim/vimrc"
>      user vimrc file: "$HOME/.vimrc"
>  2nd user vimrc file: "~/.vim/vimrc"
>       user exrc file: "$HOME/.exrc"
>   system gvimrc file: "/usr/local/etc/vim/gvimrc"
>     user gvimrc file: "$HOME/.gvimrc"
> 2nd user gvimrc file: "~/.vim/gvimrc"
>     system menu file: "$VIMRUNTIME/menu.vim"
>   fall-back for $VIM: "/usr/local/etc/vim"
>  f-b for $VIMRUNTIME: "/usr/local/share/vim/vim74"
> Compilation: cc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -I/usr/local/include/gtk-2.0 -I/usr/local/include/pango-1.0 -I/usr/local/include/glib-2.0 -I/usr/local/lib/glib-2.0/include -I/usr/local/include -I/usr/local/include/cairo -I/usr/local/include/pixman-1 -I/usr/local/include/freetype2    -I/usr/local/include/freetype2 -I/usr/local/include/libdrm -I/usr/local/include/libpng16 -I/usr/local/include/harfbuzz -I/usr/local/include/gdk-pixbuf-2.0 -I/usr/local/include/atk-1.0 -D_THREAD_SAFE -pthread -DLIBICONV_PLUG -I/usr/local/include  -O2 -pipe  -DLIBICONV_PLUG -fstack-protector -fno-strict-aliasing   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
> Linking: cc -L/usr/local/lib -lX11 -lXpm -lXt -L.  -L/usr/local/lib  -fstack-protector -rdynamic -Wl,-R/usr/local/lib/perl5/5.20/mach/CORE   -Wl,-rpath=/usr/lib:/usr/local/lib -fstack-protector -L/usr/local/lib -Wl,--as-needed -o vim     -L/usr/local/lib -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lpthread -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lintl -lfontconfig -lfreetype  -lSM -lICE -lXpm  -lX11 -lXpm -lXt  -lXdmcp -lSM -lICE    -lm -lelf  -pthread -ltermlib  -lintl      -Wl,-R/usr/local/lib/perl5/5.20/mach/CORE -lpthread -Wl,-E  -fstack-protector -L/usr/local/lib  -L/usr/local/lib/perl5/5.20/mach/CORE -lperl -lpthread -lm -lcrypt -lutil  -L/usr/local/lib/python2.7/config -lpython2.7 -L/usr/local/lib -lintl -lutil -lm -Wl,--export-dynamic    -L/usr/local/lib -ltcl86 -lz -lpthread -lm  -lruby22 -lexecinfo -lprocstat -lcrypt -lm -lpthread -L/usr/local/lib -L/usr/local/lib

- Operating system/version: FreeBSD 10.1
- Terminal name/version: putty and teraterm
- `$TERM`: xterm-256color

### Actual behaviour
HOME and END keys are swallowed before they reach nvim. Ctrl+V followed by HOME or END continues to wait for a key press (i.e. not a key mapping issue, the keys themselves are completely swallowed).

HOME and END work fine in both VIM over SSH in putty or teraterm. HOME and END keys work fine in neovim from an actual console.

In a bash prompt, pressing Ctrl-V followed by HOME prints `^[[1~` and followed by END prints `^[[4~`

### Expected behaviour

HOME and END should be mappable, preferably working out of the box without remapping. 

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
type something. Press HOME. Press END. Nothing happens.

```

**Comment by justinmk** [comment]
Same as https://github.com/neovim/neovim/issues/4809, but this has a bit more info.

**Comment by snoack** [comment]
I have the same problem, regardless of SSH, when using [Alacritty](https://github.com/jwilm/alacritty) (GPU-accelerated terminal emulator). I'm using nvim 0.1.7. When I switch either to VIM 8.0 or to a different terminal (xterm, urxvt, gnome-terminal) it works. Just the combination of Alacritty+nvim causes Home/End buttons to not work.

**Comment by justinmk** [comment]
@snoack Check `$TERM`.

**Comment by snoack** [comment]
`TERM` is set to `xterm-256color` in Alacritty. I tried changing it to just `xterm` and `rxvt-unicode`, but the issue persists.

**Comment by tex** [comment]
Setting TERM to putty-256color does make Home and End work to work. Ctrl-Home, Ctrl-End does not work.

**Comment by justinmk** [comment]
Sounds like `TERM=putty-256color` works now, thanks to @jdebp . Details are at `:help $TERM` (starting with nvim 0.2.1)

>  Ctrl-Home, Ctrl-End does not work.

Do those work in Vim (not Nvim) or any other terminal application?

**Comment by tex** [comment]
Ctrl-Home, Ctrl-End does not work in vim / nvim.
I do not use any other terminal application that would use those keys though :-)
Added the following mapping into Alacritty and it works in nvim. It does not work in vim.

  - { key: Home,     mods: Control, chars: "\x1b[1;5H",mode: ~AppCursor  }
  - { key: End,      mods: Control, chars: "\x1b[1;5F",mode: ~AppCursor  }

**Comment by petr-fischer** [comment]
I have NVIM v0.2.2 and HOME and KEY doesn't work over putty and also mobile JuiceSSH. Without SSH, HOME & END is OK. My TERM=xterm-256color (screen-256color also broken). Any ideas?

**Comment by justinmk** [comment]
@petr-fischer The terminfo on the remote system corresponding to xterm-256color is missing the correct entries for your terminal (which is putty, not xterm ...).

**Comment by Magicrafter13** [comment]
My `$TERM` variable is set to `xterm-256color` but Home and End don't work. PgUp and PgDn are fine.

I'm connecting over ssh (OpenSSH) through Powershell 7.x. With the latest version of Neovim available through apt for Ubuntu 20.04 Server.

Bash gives me the same output as @mqudsi with ^V +Home and ^V+End

**Comment by erw7** [comment]
@Magicrafter13 `C:\Windows\System32\OpenSSH\ssh.exe`, those keys don't work, which is a problem with the OpenSSH that came with Windows. Using v8.1.0.0p1-Beta on the following page should fix the problem if you have recent Windows 10.

https://github.com/PowerShell/Win32-OpenSSH/releases
